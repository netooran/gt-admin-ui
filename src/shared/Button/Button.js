import PropTypes from 'prop-types';
import styled, { css } from 'styled-components';

const defaultStyles = css`
  --button-primary-color: #000000;

  border: none;
  cursor: pointer;
  opacity: 0.7;
  font-weight: bold;
  padding: 0.5rem 1.5rem;
  border-radius: 3rem;

  svg {
    width: 1.2rem;
  }
`;

const disabledStyles = css`
  :disabled {
    opacity: 0.4;
    pointer-events: none;
  }
`;

const hoverStyles = css`
  :hover {
    opacity: 1;
  }
`;

const dangerStyles = css`
  --button-primary-color: red;
`;

const solidStyles = css`
  background-color: var(--button-primary-color);
  color: white;
`;

const outlineStyles = css`
  background-color: transparent;
  border: 1px solid var(--button-primary-color);
  color: var(--button-primary-color); ;
`;

const ghostStyles = css`
  background-color: transparent;
  color: var(--button-primary-color);
`;

const iconStyles = css`
  background-color: transparent;
  color: var(--button-primary-color);
  padding: 0;
`;

const roundStyles = css`
  min-width: 3rem;
  min-height: 3rem;
  padding: 0;
`;

const StyledButton = styled.button`
  ${defaultStyles}
  ${disabledStyles}
  ${hoverStyles}

  ${({ importance }) => importance === 'danger' && dangerStyles}
  ${({ variant }) => variant === 'solid' && solidStyles}
  ${({ variant }) => variant === 'ghost' && ghostStyles}
  ${({ variant }) => variant === 'outline' && outlineStyles}
  ${({ variant }) => variant === 'icon' && iconStyles}
  ${({ shape }) => shape === 'round' && roundStyles}
`;

const Button = ({ children, variant = 'solid', ...props }) => {
  return (
    <StyledButton variant={variant} {...props}>
      {children}
    </StyledButton>
  );
};

Button.propTypes = {
  children: PropTypes.node,
  variant: PropTypes.oneOf(['solid', 'outline', 'ghost', 'icon']),
  importance: PropTypes.oneOf(['danger']),
  shape: PropTypes.oneOf(['round']),
  disabled: PropTypes.bool,
};

export default Button;
