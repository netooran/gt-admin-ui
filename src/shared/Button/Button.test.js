import { fireEvent, render, screen } from '@testing-library/react';
import { Home } from '@styled-icons/boxicons-regular';
import Button from './Button';

describe('Default Button', () => {
  test('renders button with default props', () => {
    const onClick = jest.fn();
    const { container } = render(<Button onClick={onClick}>Click me</Button>);
    expect(container).toMatchSnapshot();

    fireEvent.click(screen.getByText('Click me'));
    expect(onClick).toHaveBeenCalled();
  });
});

describe('Solid Button', () => {
  test('renders button with primary color as background & white font', () => {
    const props = { variant: 'solid' };
    const { container } = render(<Button {...props}>Click me</Button>);
    expect(container).toMatchSnapshot();
  });

  test('renders button with red background color & white font', () => {
    const props = { variant: 'solid', importance: 'danger' };
    const { container } = render(<Button {...props}>Click me</Button>);
    expect(container).toMatchSnapshot();
  });

  test('renders button with primary color as background & white font & rounded border', () => {
    const props = { variant: 'solid', shape: 'round' };
    const { container } = render(<Button {...props}>Click me</Button>);
    expect(container).toMatchSnapshot();
  });

  test('renders button with primary color as background & white font but disabled', () => {
    const props = { variant: 'solid', disabled: true };
    const { container } = render(<Button {...props}>Click me</Button>);
    expect(container).toMatchSnapshot();
  });
});

describe('Ghost Button', () => {
  test('renders button with primary color for text & without background & border', () => {
    const props = { variant: 'ghost' };
    const { container } = render(<Button {...props}>Click me</Button>);
    expect(container).toMatchSnapshot();
  });

  test('renders button with red color for text & without background & border', () => {
    const props = { variant: 'ghost', importance: 'danger' };
    const { container } = render(<Button {...props}>Click me</Button>);
    expect(container).toMatchSnapshot();
  });

  test('renders button with primary color for text & without background & border in round shape', () => {
    const props = { variant: 'ghost', shape: 'round' };
    const { container } = render(<Button {...props}>Click me</Button>);
    expect(container).toMatchSnapshot();
  });

  test('renders button with primary color for text & without background & border but disabled', () => {
    const props = { variant: 'ghost', disabled: true };
    const { container } = render(<Button {...props}>Click me</Button>);
    expect(container).toMatchSnapshot();
  });
});

describe('Outlined Button', () => {
  test('renders button with transparent background & primary colored border and text', () => {
    const props = { variant: 'outline' };
    const { container } = render(<Button {...props}>Click me</Button>);
    expect(container).toMatchSnapshot();
  });

  test('renders button with transparent background & red colored border and text', () => {
    const props = { variant: 'outline', importance: 'danger' };
    const { container } = render(<Button {...props}>Click me</Button>);
    expect(container).toMatchSnapshot();
  });

  test('renders button with transparent background & primary colored border and text and rounded borders', () => {
    const props = { variant: 'outline', shape: 'round' };
    const { container } = render(<Button {...props}>Click me</Button>);
    expect(container).toMatchSnapshot();
  });

  test('renders button with transparent background & primary colored border and text but disabled', () => {
    const props = { variant: 'outline', disabled: true };
    const { container } = render(<Button {...props}>Click me</Button>);
    expect(container).toMatchSnapshot();
  });
});

describe('Icon Button', () => {
  test('renders button primary colored icon', () => {
    const props = { variant: 'icon' };
    const { container } = render(
      <Button {...props}>
        <Home />
      </Button>
    );
    expect(container).toMatchSnapshot();
  });

  test('renders button red colored icon', () => {
    const props = { variant: 'icon', importance: 'danger' };
    const { container } = render(
      <Button {...props}>
        <Home />
      </Button>
    );
    expect(container).toMatchSnapshot();
  });

  test('renders button primary colored icon but disabled', () => {
    const props = { variant: 'icon', disabled: true };
    const { container } = render(
      <Button {...props}>
        <Home />
      </Button>
    );
    expect(container).toMatchSnapshot();
  });
});
