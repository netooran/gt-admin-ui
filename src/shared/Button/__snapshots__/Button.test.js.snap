// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Default Button renders button with default props 1`] = `
.c0 {
  border: none;
  cursor: pointer;
  opacity: 0.7;
  font-weight: bold;
  padding: 0.5rem 1.5rem;
  border-radius: 3rem;
  --button-primary-color: #000000;
  background-color: var(--button-primary-color);
  color: white;
}

.c0 svg {
  width: 1.2rem;
}

.c0:hover {
  opacity: 1;
}

<div>
  <button
    class="c0"
  >
    Click me
  </button>
</div>
`;

exports[`Ghost Button renders button with primary color for text & without background & border 1`] = `
.c0 {
  border: none;
  cursor: pointer;
  opacity: 0.7;
  font-weight: bold;
  padding: 0.5rem 1.5rem;
  border-radius: 3rem;
  --button-primary-color: #000000;
  background-color: transparent;
  color: var(--button-primary-color);
}

.c0 svg {
  width: 1.2rem;
}

.c0:hover {
  opacity: 1;
}

<div>
  <button
    class="c0"
  >
    Click me
  </button>
</div>
`;

exports[`Ghost Button renders button with primary color for text & without background & border but disabled 1`] = `
.c0 {
  border: none;
  cursor: pointer;
  opacity: 0.7;
  font-weight: bold;
  padding: 0.5rem 1.5rem;
  border-radius: 3rem;
  --button-primary-color: #000000;
  background-color: transparent;
  color: var(--button-primary-color);
  opacity: 0.4;
  pointer-events: none;
}

.c0 svg {
  width: 1.2rem;
}

.c0:hover {
  opacity: 1;
}

<div>
  <button
    class="c0"
    disabled=""
  >
    Click me
  </button>
</div>
`;

exports[`Ghost Button renders button with primary color for text & without background & border in round shape 1`] = `
.c0 {
  border: none;
  cursor: pointer;
  opacity: 0.7;
  font-weight: bold;
  padding: 0.5rem 1.5rem;
  border-radius: 3rem;
  --button-primary-color: #000000;
  background-color: transparent;
  color: var(--button-primary-color);
  min-width: 3rem;
  min-height: 3rem;
  padding: 0;
}

.c0 svg {
  width: 1.2rem;
}

.c0:hover {
  opacity: 1;
}

<div>
  <button
    class="c0"
    shape="round"
  >
    Click me
  </button>
</div>
`;

exports[`Ghost Button renders button with red color for text & without background & border 1`] = `
.c0 {
  border: none;
  cursor: pointer;
  opacity: 0.7;
  font-weight: bold;
  padding: 0.5rem 1.5rem;
  border-radius: 3rem;
  --button-primary-color: #000000;
  --button-primary-color: red;
  background-color: transparent;
  color: var(--button-primary-color);
}

.c0 svg {
  width: 1.2rem;
}

.c0:hover {
  opacity: 1;
}

<div>
  <button
    class="c0"
  >
    Click me
  </button>
</div>
`;

exports[`Icon Button renders button primary colored icon 1`] = `
.c1 {
  display: inline-block;
  vertical-align: middle;
  overflow: hidden;
}

.c0 {
  border: none;
  cursor: pointer;
  opacity: 0.7;
  font-weight: bold;
  padding: 0.5rem 1.5rem;
  border-radius: 3rem;
  --button-primary-color: #000000;
  background-color: transparent;
  color: var(--button-primary-color);
  padding: 0;
}

.c0 svg {
  width: 1.2rem;
}

.c0:hover {
  opacity: 1;
}

<div>
  <button
    class="c0"
  >
    <svg
      aria-hidden="true"
      class="c1"
      fill="currentColor"
      focusable="false"
      viewBox="0 0 24 24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M3 13h1v7c0 1.103.897 2 2 2h12c1.103 0 2-.897 2-2v-7h1a1 1 0 00.707-1.707l-9-9a.999.999 0 00-1.414 0l-9 9A1 1 0 003 13zm7 7v-5h4v5h-4zm2-15.586l6 6V15l.001 5H16v-5c0-1.103-.897-2-2-2h-4c-1.103 0-2 .897-2 2v5H6v-9.586l6-6z"
      />
    </svg>
  </button>
</div>
`;

exports[`Icon Button renders button primary colored icon but disabled 1`] = `
.c1 {
  display: inline-block;
  vertical-align: middle;
  overflow: hidden;
}

.c0 {
  border: none;
  cursor: pointer;
  opacity: 0.7;
  font-weight: bold;
  padding: 0.5rem 1.5rem;
  border-radius: 3rem;
  --button-primary-color: #000000;
  background-color: transparent;
  color: var(--button-primary-color);
  padding: 0;
  opacity: 0.4;
  pointer-events: none;
}

.c0 svg {
  width: 1.2rem;
}

.c0:hover {
  opacity: 1;
}

<div>
  <button
    class="c0"
    disabled=""
  >
    <svg
      aria-hidden="true"
      class="c1"
      fill="currentColor"
      focusable="false"
      viewBox="0 0 24 24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M3 13h1v7c0 1.103.897 2 2 2h12c1.103 0 2-.897 2-2v-7h1a1 1 0 00.707-1.707l-9-9a.999.999 0 00-1.414 0l-9 9A1 1 0 003 13zm7 7v-5h4v5h-4zm2-15.586l6 6V15l.001 5H16v-5c0-1.103-.897-2-2-2h-4c-1.103 0-2 .897-2 2v5H6v-9.586l6-6z"
      />
    </svg>
  </button>
</div>
`;

exports[`Icon Button renders button red colored icon 1`] = `
.c1 {
  display: inline-block;
  vertical-align: middle;
  overflow: hidden;
}

.c0 {
  border: none;
  cursor: pointer;
  opacity: 0.7;
  font-weight: bold;
  padding: 0.5rem 1.5rem;
  border-radius: 3rem;
  --button-primary-color: #000000;
  --button-primary-color: red;
  background-color: transparent;
  color: var(--button-primary-color);
  padding: 0;
}

.c0 svg {
  width: 1.2rem;
}

.c0:hover {
  opacity: 1;
}

<div>
  <button
    class="c0"
  >
    <svg
      aria-hidden="true"
      class="c1"
      fill="currentColor"
      focusable="false"
      viewBox="0 0 24 24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M3 13h1v7c0 1.103.897 2 2 2h12c1.103 0 2-.897 2-2v-7h1a1 1 0 00.707-1.707l-9-9a.999.999 0 00-1.414 0l-9 9A1 1 0 003 13zm7 7v-5h4v5h-4zm2-15.586l6 6V15l.001 5H16v-5c0-1.103-.897-2-2-2h-4c-1.103 0-2 .897-2 2v5H6v-9.586l6-6z"
      />
    </svg>
  </button>
</div>
`;

exports[`Outlined Button renders button with transparent background & primary colored border and text 1`] = `
.c0 {
  border: none;
  cursor: pointer;
  opacity: 0.7;
  font-weight: bold;
  padding: 0.5rem 1.5rem;
  border-radius: 3rem;
  --button-primary-color: #000000;
  background-color: transparent;
  border: 1px solid var(--button-primary-color);
  color: var(--button-primary-color);
}

.c0 svg {
  width: 1.2rem;
}

.c0:hover {
  opacity: 1;
}

<div>
  <button
    class="c0"
  >
    Click me
  </button>
</div>
`;

exports[`Outlined Button renders button with transparent background & primary colored border and text and rounded borders 1`] = `
.c0 {
  border: none;
  cursor: pointer;
  opacity: 0.7;
  font-weight: bold;
  padding: 0.5rem 1.5rem;
  border-radius: 3rem;
  --button-primary-color: #000000;
  background-color: transparent;
  border: 1px solid var(--button-primary-color);
  color: var(--button-primary-color);
  min-width: 3rem;
  min-height: 3rem;
  padding: 0;
}

.c0 svg {
  width: 1.2rem;
}

.c0:hover {
  opacity: 1;
}

<div>
  <button
    class="c0"
    shape="round"
  >
    Click me
  </button>
</div>
`;

exports[`Outlined Button renders button with transparent background & primary colored border and text but disabled 1`] = `
.c0 {
  border: none;
  cursor: pointer;
  opacity: 0.7;
  font-weight: bold;
  padding: 0.5rem 1.5rem;
  border-radius: 3rem;
  --button-primary-color: #000000;
  background-color: transparent;
  border: 1px solid var(--button-primary-color);
  color: var(--button-primary-color);
  opacity: 0.4;
  pointer-events: none;
}

.c0 svg {
  width: 1.2rem;
}

.c0:hover {
  opacity: 1;
}

<div>
  <button
    class="c0"
    disabled=""
  >
    Click me
  </button>
</div>
`;

exports[`Outlined Button renders button with transparent background & red colored border and text 1`] = `
.c0 {
  border: none;
  cursor: pointer;
  opacity: 0.7;
  font-weight: bold;
  padding: 0.5rem 1.5rem;
  border-radius: 3rem;
  --button-primary-color: #000000;
  --button-primary-color: red;
  background-color: transparent;
  border: 1px solid var(--button-primary-color);
  color: var(--button-primary-color);
}

.c0 svg {
  width: 1.2rem;
}

.c0:hover {
  opacity: 1;
}

<div>
  <button
    class="c0"
  >
    Click me
  </button>
</div>
`;

exports[`Solid Button renders button with primary color as background & white font & rounded border 1`] = `
.c0 {
  border: none;
  cursor: pointer;
  opacity: 0.7;
  font-weight: bold;
  padding: 0.5rem 1.5rem;
  border-radius: 3rem;
  --button-primary-color: #000000;
  background-color: var(--button-primary-color);
  color: white;
  min-width: 3rem;
  min-height: 3rem;
  padding: 0;
}

.c0 svg {
  width: 1.2rem;
}

.c0:hover {
  opacity: 1;
}

<div>
  <button
    class="c0"
    shape="round"
  >
    Click me
  </button>
</div>
`;

exports[`Solid Button renders button with primary color as background & white font 1`] = `
.c0 {
  border: none;
  cursor: pointer;
  opacity: 0.7;
  font-weight: bold;
  padding: 0.5rem 1.5rem;
  border-radius: 3rem;
  --button-primary-color: #000000;
  background-color: var(--button-primary-color);
  color: white;
}

.c0 svg {
  width: 1.2rem;
}

.c0:hover {
  opacity: 1;
}

<div>
  <button
    class="c0"
  >
    Click me
  </button>
</div>
`;

exports[`Solid Button renders button with primary color as background & white font but disabled 1`] = `
.c0 {
  border: none;
  cursor: pointer;
  opacity: 0.7;
  font-weight: bold;
  padding: 0.5rem 1.5rem;
  border-radius: 3rem;
  --button-primary-color: #000000;
  background-color: var(--button-primary-color);
  color: white;
  opacity: 0.4;
  pointer-events: none;
}

.c0 svg {
  width: 1.2rem;
}

.c0:hover {
  opacity: 1;
}

<div>
  <button
    class="c0"
    disabled=""
  >
    Click me
  </button>
</div>
`;

exports[`Solid Button renders button with red background color & white font 1`] = `
.c0 {
  border: none;
  cursor: pointer;
  opacity: 0.7;
  font-weight: bold;
  padding: 0.5rem 1.5rem;
  border-radius: 3rem;
  --button-primary-color: #000000;
  --button-primary-color: red;
  background-color: var(--button-primary-color);
  color: white;
}

.c0 svg {
  width: 1.2rem;
}

.c0:hover {
  opacity: 1;
}

<div>
  <button
    class="c0"
  >
    Click me
  </button>
</div>
`;
